File: app/src/main/java/com/asfoundation/wallet/ui/iab/AppcoinsRewardsBuyFragment.java
package com.asfoundation.wallet.ui.iab;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.airbnb.lottie.LottieAnimationView;
import com.appcoins.wallet.bdsbilling.repository.entity.Purchase;
import com.appcoins.wallet.billing.BillingMessagesMapper;
import com.asf.wallet.R;
import com.asfoundation.wallet.billing.analytics.BillingAnalytics;
import com.asfoundation.wallet.entity.TransactionBuilder;
import com.asfoundation.wallet.util.CurrencyFormatUtils;
import com.asfoundation.wallet.util.TransferParser;
import com.jakewharton.rxbinding2.view.RxView;
import dagger.android.support.DaggerFragment;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import java.math.BigDecimal;
import javax.inject.Inject;

import static com.asfoundation.wallet.ui.iab.InAppPurchaseInteractor.PRE_SELECTED_PAYMENT_METHOD_KEY;

public class AppcoinsRewardsBuyFragment extends DaggerFragment implements AppcoinsRewardsBuyView {

  private static final String PRODUCT_NAME = "product_name";
  private static final String AMOUNT_KEY = "amount";
  private static final String URI_KEY = "uri_key";
  private static final String IS_BDS = "is_bds";
  private static final String TRANSACTION_KEY = "transaction_key";
  private static final String GAMIFICATION_LEVEL = "gamification_level";

  @Inject RewardsManager rewardsManager;
  @Inject TransferParser transferParser;
  @Inject BillingMessagesMapper billingMessagesMapper;
  @Inject BillingAnalytics analytics;
  @Inject CurrencyFormatUtils formatter;
  @Inject AppcoinsRewardsBuyInteract appcoinsRewardsBuyInteract;
  private View loadingView;
  private View transactionCompletedLayout;
  private LottieAnimationView lottieTransactionComplete;
  private AppcoinsRewardsBuyPresenter presenter;
  private BigDecimal amount;
  private IabView iabView;
  private String uri;
  private boolean isBds;
  private View transactionErrorLayout;
  private TextView errorMessage;
  private Button okErrorButton;
  private View supportIcon;
  private View supportLogo;
  private int gamificationLevel;

  public static Fragment newInstance(BigDecimal amount, TransactionBuilder transactionBuilder,
      String uri, String productName, boolean isBds, int gamificationLevel) {
    AppcoinsRewardsBuyFragment fragment = new AppcoinsRewardsBuyFragment();
    Bundle bundle = new Bundle();
    bundle.putString(AMOUNT_KEY, amount.toString());
    bundle.putParcelable(TRANSACTION_KEY, transactionBuilder);
    bundle.putString(URI_KEY, uri);
    bundle.putString(PRODUCT_NAME, productName);
    bundle.putBoolean(IS_BDS, isBds);
    bundle.putInt(GAMIFICATION_LEVEL, gamificationLevel);
    fragment.setArguments(bundle);
    return fragment;
  }

  @Override public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Bundle arguments = getArguments();
    amount = new BigDecimal(arguments.getString(AMOUNT_KEY));
    uri = arguments.getString(URI_KEY);
    isBds = arguments.getBoolean(IS_BDS);
    gamificationLevel = arguments.getInt(GAMIFICATION_LEVEL);
  }

  @Nullable @Override
  public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    return inflater.inflate(R.layout.reward_payment_layout, container, false);
  }

  @Override public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    loadingView = view.findViewById(R.id.loading_view);

    errorMessage = view.findViewById(R.id.error_message);
    transactionErrorLayout = view.findViewById(R.id.generic_error_layout);
    okErrorButton = view.findViewById(R.id.error_dismiss);
    supportIcon = view.findViewById(R.id.layout_support_icn);
    supportLogo = view.findViewById(R.id.layout_support_logo);
    transactionCompletedLayout = view.findViewById(R.id.iab_activity_transaction_completed);

    TransactionBuilder transactionBuilder = getArguments().getParcelable(TRANSACTION_KEY);
    String callerPackageName = transactionBuilder.getDomain();
    presenter =
        new AppcoinsRewardsBuyPresenter(this, rewardsManager, AndroidSchedulers.mainThread(),
            new CompositeDisposable(), amount, uri, callerPackageName, transferParser, isBds,
            analytics, transactionBuilder, formatter, gamificationLevel,
            appcoinsRewardsBuyInteract);

    lottieTransactionComplete =
        transactionCompletedLayout.findViewById(R.id.lottie_transaction_success);

    setupTransactionCompleteAnimation();

    presenter.present();
  }

  @Override public void onDestroyView() {
    presenter.stop();

    super.onDestroyView();
  }

  @Override public void finish(Purchase purchase) {
    finish(purchase, null);
  }

  @Override public void showLoading() {
    transactionErrorLayout.setVisibility(View.GONE);
    transactionCompletedLayout.setVisibility(View.INVISIBLE);
    loadingView.setVisibility(View.VISIBLE);
  }

  @Override public void hideLoading() {
    loadingView.setVisibility(View.GONE);
  }

  @Override public void showNoNetworkError() {
    hideLoading();
    okErrorButton.setText(R.string.ok);
    errorMessage.setText(R.string.activity_iab_no_network_message);
    transactionErrorLayout.setVisibility(View.VISIBLE);
  }

  @Override public Observable<Object> getOkErrorClick() {
    return RxView.clicks(okErrorButton);
  }

  @Override public Observable<Object> getSupportIconClick() {
    return RxView.clicks(supportIcon);
  }

  @Override public Observable<Object> getSupportLogoClick() {
    return RxView.clicks(supportLogo);
  }

  @Override public void close() {
    iabView.close(billingMessagesMapper.mapCancellation());
  }

  @Override public void showGenericError() {
    showError(null);
  }

  @Override public void showError(Integer message) {
    okErrorButton.setText(R.string.ok);
    errorMessage.setText(
        getString(message != null ? message : R.string.activity_iab_error_message));
    transactionErrorLayout.setVisibility(View.VISIBLE);
    hideLoading();
  }

  @Override public void finish(String uid) {
    presenter.sendPaymentEvent();
    presenter.sendRevenueEvent();
    presenter.sendPaymentSuccessEvent();
    Bundle bundle = billingMessagesMapper.successBundle(uid);
    bundle.putString(PRE_SELECTED_PAYMENT_METHOD_KEY,
        PaymentMethodsView.PaymentMethodId.APPC_CREDITS.getId());
    iabView.finish(bundle);
  }

  @Override public void errorClose() {
    iabView.close(billingMessagesMapper.genericError());
  }

  @Override public void finish(Purchase purchase, @Nullable String orderReference) {
    presenter.sendPaymentEvent();
    presenter.sendRevenueEvent();
    presenter.sendPaymentSuccessEvent();
    Bundle bundle = billingMessagesMapper.mapPurchase(purchase, orderReference);
    bundle.putString(PRE_SELECTED_PAYMENT_METHOD_KEY,
        PaymentMethodsView.PaymentMethodId.APPC_CREDITS.getId());
    iabView.finish(bundle);
  }

  @Override public void showTransactionCompleted() {
    loadingView.setVisibility(View.GONE);
    transactionErrorLayout.setVisibility(View.GONE);
    transactionCompletedLayout.setVisibility(View.VISIBLE);
  }

  @Override public long getAnimationDuration() {
    return lottieTransactionComplete.getDuration();
  }

  @Override public void lockRotation() {
    iabView.lockRotation();
  }

  @Override public void onAttach(Context context) {
    super.onAttach(context);
    if (!(context instanceof IabView)) {
      throw new IllegalStateException(
          "Express checkout buy fragment must be attached to IAB activity");
    }
    iabView = ((IabView) context);
  }

  private void setupTransactionCompleteAnimation() {
    LottieAnimationView lottieTransactionComplete =
        transactionCompletedLayout.findViewById(R.id.lottie_transaction_success);
    lottieTransactionComplete.setAnimation(R.raw.success_animation);
  }
}

File: app/src/main/java/com/asfoundation/wallet/ui/iab/AppcoinsRewardsBuyFragment.kt
package com.asfoundation.wallet.ui.iab

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.appcoins.wallet.bdsbilling.repository.entity.Purchase
import com.appcoins.wallet.billing.BillingMessagesMapper
import com.appcoins.wallet.core.analytics.analytics.legacy.BillingAnalytics
import com.appcoins.wallet.core.utils.android_common.CurrencyFormatUtils
import com.appcoins.wallet.core.utils.jvm_common.Logger
import com.asf.wallet.R
import com.asf.wallet.databinding.RewardPaymentLayoutBinding
import com.asfoundation.wallet.entity.TransactionBuilder
import com.asfoundation.wallet.util.TransferParser
import com.jakewharton.rxbinding2.view.RxView
import com.wallet.appcoins.core.legacy_base.BasePageViewFragment
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import java.math.BigDecimal
import javax.inject.Inject

@AndroidEntryPoint
class AppcoinsRewardsBuyFragment : BasePageViewFragment(), AppcoinsRewardsBuyView {

  @Inject
  lateinit var rewardsManager: RewardsManager

  @Inject
  lateinit var transferParser: TransferParser

  @Inject
  lateinit var billingMessagesMapper: BillingMessagesMapper

  @Inject
  lateinit var analytics: BillingAnalytics

  @Inject
  lateinit var paymentAnalytics: PaymentMethodsAnalytics

  @Inject
  lateinit var formatter: CurrencyFormatUtils

  @Inject
  lateinit var appcoinsRewardsBuyInteract: AppcoinsRewardsBuyInteract

  @Inject
  lateinit var logger: Logger

  private lateinit var presenter: AppcoinsRewardsBuyPresenter
  private lateinit var iabView: IabView

  private val binding by lazy { RewardPaymentLayoutBinding.bind(requireView()) }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View = RewardPaymentLayoutBinding.inflate(inflater).root

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    presenter = AppcoinsRewardsBuyPresenter(
      view = this,
      rewardsManager = rewardsManager,
      viewScheduler = AndroidSchedulers.mainThread(),
      networkScheduler = Schedulers.io(),
      disposables = CompositeDisposable(),
      packageName = transactionBuilder.domain,
      isBds = isBds,
      isPreSelected = isPreSelected,
      analytics = analytics,
      paymentAnalytics = paymentAnalytics,
      transactionBuilder = transactionBuilder,
      formatter = formatter,
      gamificationLevel = gamificationLevel,
      appcoinsRewardsBuyInteract = appcoinsRewardsBuyInteract,
      logger = logger
    )
    setupTransactionCompleteAnimation()
    presenter.present()
  }

  override fun onDestroyView() {
    presenter.stop()
    super.onDestroyView()
  }

  override fun finish(purchase: Purchase) = finish(purchase, null)

  override fun showLoading() {
    binding.genericErrorLayout.genericPurchaseErrorLayout.visibility = View.GONE
    binding.fragmentIabTransactionCompleted.iabActivityTransactionCompleted.visibility = View.INVISIBLE
    binding.loadingAnimation.visibility = View.VISIBLE
    binding.makingPurchaseText.visibility = View.VISIBLE
  }

  override fun hideLoading() {
    binding.loadingAnimation.visibility = View.GONE
    binding.makingPurchaseText.visibility = View.GONE
  }

  override fun showNoNetworkError() {
    hideLoading()
    binding.genericErrorLayout.errorDismiss.setText(getString(R.string.ok))
    binding.genericErrorLayout.genericErrorLayout.errorMessage.setText(R.string.activity_iab_no_network_message)
    binding.genericErrorLayout.genericPurchaseErrorLayout.visibility = View.VISIBLE
  }

  override fun getOkErrorClick() = RxView.clicks(binding.genericErrorLayout.errorDismiss)

  override fun getSupportIconClick() = RxView.clicks(binding.genericErrorLayout.genericErrorLayout.layoutSupportIcn)

  override fun getSupportLogoClick() = RxView.clicks(binding.genericErrorLayout.genericErrorLayout.layoutSupportLogo)

  override fun close() = iabView.close(billingMessagesMapper.mapCancellation())

  override fun showError(message: Int?) {
    binding.genericErrorLayout.errorDismiss.setText(getString(R.string.back_button))
    binding.genericErrorLayout.genericErrorLayout.errorMessage.text = getString(message ?: R.string.activity_iab_error_message)
    binding.genericErrorLayout.genericPurchaseErrorLayout.visibility = View.VISIBLE
    hideLoading()
  }

  override fun finish(uid: String?, purchaseUid: String) {
    presenter.sendPaymentEvent()
    presenter.sendRevenueEvent()
    presenter.sendPaymentSuccessEvent(purchaseUid)
    val bundle = billingMessagesMapper.successBundle(uid)
    bundle.putString(
      InAppPurchaseInteractor.PRE_SELECTED_PAYMENT_METHOD_KEY,
      PaymentMethodsView.PaymentMethodId.APPC_CREDITS.id
    )
    iabView.finish(bundle)
  }

  override fun errorClose() = iabView.close(billingMessagesMapper.genericError())

  override fun showPaymentMethods() {
    iabView.unlockRotation()
    iabView.showPaymentMethodsView()
  }

  override fun finish(purchase: Purchase, orderReference: String?) {
    presenter.sendPaymentEvent()
    presenter.sendRevenueEvent()
    presenter.sendPaymentSuccessEvent(purchase.uid)
    val bundle = billingMessagesMapper.mapPurchase(purchase, orderReference)
    bundle.putString(
      InAppPurchaseInteractor.PRE_SELECTED_PAYMENT_METHOD_KEY,
      PaymentMethodsView.PaymentMethodId.APPC_CREDITS.id
    )
    iabView.finish(bundle)
  }

  override fun showVerification() = iabView.showVerification(false)

  override fun showTransactionCompleted() {
    binding.loadingAnimation.visibility = View.GONE
    binding.makingPurchaseText.visibility = View.GONE
    binding.genericErrorLayout.genericPurchaseErrorLayout.visibility = View.GONE
    binding.fragmentIabTransactionCompleted.iabActivityTransactionCompleted.visibility = View.VISIBLE
    binding.fragmentIabTransactionCompleted.bonusSuccessLayout.visibility = View.GONE
  }

  override fun getAnimationDuration() = binding.fragmentIabTransactionCompleted.lottieTransactionSuccess.duration * 2

  override fun lockRotation() = iabView.lockRotation()

  override fun onAttach(context: Context) {
    super.onAttach(context)
    check(context is IabView) { "AppcoinsRewardsBuyFragment must be attached to IAB activity" }
    iabView = context
  }

  private fun setupTransactionCompleteAnimation() =
    binding.fragmentIabTransactionCompleted.lottieTransactionSuccess.setAnimation(R.raw.success_animation)

  private val isBds: Boolean by lazy {
    if (requireArguments().containsKey(IS_BDS)) {
      requireArguments().getBoolean(IS_BDS)
    } else {
      throw IllegalArgumentException("isBds not found")
    }
  }

  private val isPreSelected: Boolean by lazy {
    if (requireArguments().containsKey(PRE_SELECTED_KEY)) {
      requireArguments().getBoolean(PRE_SELECTED_KEY)
    } else {
      throw IllegalArgumentException("pre selected data not found")
    }
  }

  private val gamificationLevel: Int by lazy {
    if (requireArguments().containsKey(GAMIFICATION_LEVEL)) {
      requireArguments().getInt(GAMIFICATION_LEVEL)
    } else {
      throw IllegalArgumentException("gamification level data not found")
    }
  }

  private val transactionBuilder: TransactionBuilder by lazy {
    if (requireArguments().containsKey(TRANSACTION_KEY)) {
      requireArguments().getParcelable<TransactionBuilder>(TRANSACTION_KEY)!!
    } else {
      throw IllegalArgumentException("transaction data not found")
    }
  }

  companion object {
    private const val AMOUNT_KEY = "amount"
    private const val URI_KEY = "uri_key"
    private const val IS_BDS = "is_bds"
    private const val TRANSACTION_KEY = "transaction_key"
    private const val GAMIFICATION_LEVEL = "gamification_level"
    private const val PRE_SELECTED_KEY = "pre_selected"

    fun newInstance(
      amount: BigDecimal,
      transactionBuilder: TransactionBuilder,
      uri: String?,
      isBds: Boolean,
      isPreSelected: Boolean,
      gamificationLevel: Int
    ): Fragment = AppcoinsRewardsBuyFragment().apply {
      arguments = Bundle().apply {
        putSerializable(AMOUNT_KEY, amount)
        putParcelable(TRANSACTION_KEY, transactionBuilder)
        putString(URI_KEY, uri)
        putBoolean(IS_BDS, isBds)
        putBoolean(PRE_SELECTED_KEY, isPreSelected)
        putInt(GAMIFICATION_LEVEL, gamificationLevel)
      }
    }
  }
}